### ISSUE TITLE
Extending 'watchContracts' and killing parity periodically

### ISSUE THREAD

**Issue by tjayrush** [Great-Hill-Corporation/quickBlocks#505]
I run the Parity Ethereum node with this command line: `parity --tracing on --pruning archive`.  I then run something I call `watchContract` which is a simple shell script. Here:

    #!/usr/bin/env bash
    #
    # usage: watch [command] [sleep duration]
    #
    # Repeatedly runs the given command every duration seconds
    cd ../monitors
    while :; do
        clear
        blockScrape -fl
        blockAcct variable
        miniBlocks freshen
        time make -f freshen.mak
        echo "Sleeping for $1 seconds"
        sleep $1
    done

Parity, for some unknown reason, crashes about twice a day because it writes 400-600 GB of data to the hard drive. If you kill parity before the disc fills up and then start it again, it cleans up the mess and does not fill the disc. Some guy wrote this code (which I think is Python) to kill parity whenever the disc space grows within 90% of the hard drive space.

Here's that code:

    import subprocess
    import os
    import time
    import signal

    while True:
        print("Running Parity...")
        proc = subprocess.Popen(['parity', '--tracing', 'on', '--pruning', 'archive'])
        print("Parity running with pid {0}".format(proc.pid))
        while True:
            time.sleep(30)
            # https://stackoverflow.com/a/787832
            s = os.statvfs('/')
            gigs_left = (s.f_bavail * s.f_frsize) / 1024 / 1024 / 1024
            print('{0} GB left'.format(gigs_left))
            if gigs_left < 90:
                break
        print("Terminating Parity...")
        os.kill(proc.pid, signal.SIGINT)
        proc.wait()

I would like to combine the two programs so that the `watchContract` code works whenever parity is running, and waits until parity returns after it's killed.  Maybe this doesn't make sense, so please ask, but if possible, just post the modified code back into this issue. **Note:** you won't be able to run `watchContracts` as you don't have that code in the repo.

I need to get this working before I travel to my conference on Oct. 31.

**Comment by tjayrush** [comment]
If the above code is not Python, please port it to Python if possible. I'd rather not introduce another language into the mix.

**Comment by crodnun** [comment]
Hi Thomas, I have work a bit with the script requested but as you stated I can not fully test it at my side.
Find attached and initial draft (it is a zipped .py file)
[watchContract.py.gz](https://github.com/Great-Hill-Corporation/quickBlocks/files/1407423/watchContract.py.gz)
 
It has the following logic:



0) Its usage is xxxx.py <timeout>. i.e.  watchContract 5
1) It checks if parity is running if not, it starts it with the options '--tracing', 'on', '--pruning', 'archive'
    (feel free to modify them if you want to use the option '--warp' only.

2) It waits a guard time while parity starts given by the parameter startup_timeout (configurable)
3) Once guard time is expired, we execute every <timeout> seconds the commands you normally issue at the old bash script. Here I have a doubt with the "time make part" that is commented out.

4) After running the commands, it also checks if parity restart is required, checking the % of free space is higher than 10%. If it is lower, it restarts parity.

5) Next round of the tools will wait till parity is available again.

Let me know if this is what you were looking for, test it at your side, I could not do it here, just guess what would happen but is a beta version developed in "blind mode"  :-).

**Comment by tjayrush** [comment]
Is there any way to 'step through' a python script the way I can step through C++ code in xCode or Developer Studio?  Is there a python debugger?

**Comment by crodnun** [comment]
Yes, try with this:

https://docs.python.org/2/library/pdb.html


basically you run the script with this option:

python -m pdb myscript.py

And then you can you the common debugger statements (also described at link provided) like:

s(step): for next step
n(next): for next line
etc...

give it a look, it is all described there at section **26.3. Debugger Commands**¶

**Comment by tjayrush** [comment]
I posted the code here:

https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py

I have specific questions at specific lines:

1. Question: These commands must run synchronously (one starts after the previous one finishes) because they depend on each other. I assume these will not be parallelized, but I wanted to ask explicitly. Are these synchronous?
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py
#L28

2. Question: I will write a separate issue about this line
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L33

3. Question: We can probably comment out this line
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L40

4. Question: Good idea. Thanks.
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L55

5. Question: Should say "start running if **not** running..."
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L80

6. Question: I assume that Python's sleep command takes `seconds` as an argument here. I usually run the command with `watchContracts 20` meaning once every 20 seconds.
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L90

7. Question: I wrote an issue (#506) against this line, but I then promptly `delayed` it. The hard drive we want to watch is not necessarily the root of the drive where `watchContracts` is running. Parity can be told where to store its data, and it is the root of this disc that needs to be watched. Since Parity will hopefully one day fix their bug, we won't worry about it, but we can revisit the issue later.
   - Line: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L92

The rest of the code looks wonderful (you write very good, very easy to understand, and very well documented code--thanks for that).

You may simply answer these questions or comments by number in your response.

I will add an issue against this line now: https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/monitors/scripts/watchContract.py#L33

**Comment by crodnun** [comment]
Here you have my answers, but let's review them at meeting:

1) They you should run in order, one after other.
2) OK understood now, we need to tweak the script then to add this logic.
3) Agree, I forget to do it, it was just a debugging line, can be commented out.
4) We leave a configurable guard time while parity starts, before running our tools.
5) Correct, a typo there
6) Yes, the script uses seconds everywhere. sleep uses internally float values but for us it will be always an integer and just do the cast to float.
7) OK, no problem, if we need to tweak the filesystem logic, just let me know how you would like to proceed and I will adapt the script to use the suitable directory.
8) For the moment, everything goes to default standar output/error. It is a beta version :-) and I wanted to display everything on screen at this stage, more not being able to test everything at my side. If you consider that maybe we can move parity to a log file or maybe the commands output, I will modify the script.

I think that the best way to debug this is getting your feedback as user, and looking at the output obtained in real life, maybe redirecting whole script to a file and sending me this information. Of course first we need to add the "make" part missing (I will add it today after our meeting).

**Comment by crodnun** [comment]
Hi Thomas, the script has been updated/pushed with your comments and also added the logic to run the contracts in each directory. I need now a tester ;-) for the script. Do you know anybody to do it? :-) :-). I need your feedback regarding the same to refine/improve the same.

**Comment by tjayrush** [comment]
I can test it!!  Now it's my turn to find bugs!!

I think the code itself is exactly right, but the output doesn't work quite right.

When it runs, it says "Parity is already running" which is find, and then, I presume it goes into the watchcontract code and starts scanning.  The trouble is there is no output.

I think you may need to re-direct the standard out and the standard error. If possible redirect them to their individual places (in other words the monitor's stdout to python's stdout, and stderr to python's stderr) because I sometimes use `command..... 2>/dev/null` to capture just data to the screen.

**Comment by crodnun** [comment]
Hi Thomas, could you share somehow a version of any of these tools? (informally, I will destroy them once tested, I promise :-) :-)), otherwise it would be really time consuming to debug them if I can not see how they execute, maybe we can do this tomorrow together. With one example is enough for me to tweak the python codes, I can rename the same binary and use it everywhere (different contract names but same underlying binary).

**Comment by tjayrush** [comment]
Create a fake tool in ./quickBlocks/monitor/accts/bin/accts that just prints “Hello world” to the screen.

When you see “Hello world”, it’s working.

There’s no problem with giving you the files for the monitors, but it wouldn’t work with a lot of work—there’s a ton of hard coded nonsense that won’t port easily. 

If you can get the “Hello world” to print for one of the monitors, I think that’s all I need.

-------------------------------
Thomas Jay Rush
QuickBlocks.io
jrush@greathill.com
(610) 519-9413

**Comment by crodnun** [comment]
Yes Thomas I already tested it this way when creating the script, using a fake binary to see its output, and it seems working at my side, so maybe there is an slight different somewhere and the root cause of not displaying information on screen.
Let's review this tomorrow and together, otherwise we will waste a lot of time guessing things.

**Comment by tjayrush** [comment]
Silly me. It was sleeping for 10 seconds. It works fine.

I do have a question: You handle Control+C.  Did you do that so the script would quit if one of the called programs got a Control+C?  The reason I'm asking is because my code has to handle Control+C as well (to do some cleanup before quitting).  I'm only asking because I wanted to make sure that the watchContract.py Control+C handler doesn't interfere with the blockScrape handler---I'm 99.9% sure it won't (and it doesn't appear to from what I see), but I thought it couldn't hurt to ask.

I'm going to close this one since I can take it from here. Thanks for your help on this one. I know it was more difficult that it could have been.

**Comment by tjayrush** [comment]
I'm going to re-open because I had another question. Where is Parity running in this scenario? Does Parity run in the background? Where does parity's output go (it generates very useful logs)?  Normally, I run parity in a separate window because that way I can watch the logs. I don't suppose it's easy to run parity in another window, is it?

**Comment by crodnun** [comment]
I answer here the latest two questions:

1) Yes, no problem. If you press the Ctr+C while executing a watcher binary, it is only sent the signal to this process but the python script continues. In my fake watcher I have an sleep and a loop to simulate  it is writting many lines with some delay, and it works as I mentioned above: if we are running a child process (any quickblocks binary in this case), the Ctrl+C is sent to this process, if not is caught by the python script and stops everything.

2) Regarding parity execution, for the moment the behaviour is what I wrote you here (at the post I wrote you 2 days ago):

**For the moment, everything goes to default standar output/error. It is a beta version :-) and I wanted to display everything on screen at this stage, more not being able to test everything at my side. If you consider that maybe we can move parity to a log file or maybe the commands output, I will modify the script.**

I do not know if it is feasible to open a new console easily, but think that the solution will not be portable. What I do at my side is just to launch parity in background and redirect the output to a log file, something like this (at command line):

nohup parity --warp > /path/I/want/parity.log 2>&1

This way I can always visit the path and tail the output of the parity command.
